testing/src/test/java/io/stargate/it/driver/CqlSessionExtension.java

```diff
@@ -0,0 +1,405 @@
+      awaitAllNodes(session, stargate);
+
+      if (cqlSessionSpec.createKeyspace()) {
+        keyspaceId = generateKeyspaceId(context);
+        LOG.debug("Creating keyspace {}", keyspaceId.asCql(true));
+
+        session.execute(
+            String.format(
+                "CREATE KEYSPACE IF NOT EXISTS %s "
+                    + "WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1}",
+                keyspaceId.asCql(false)));
+        session.execute(String.format("USE %s", keyspaceId.asCql(false)));
+      }
+
+      for (String query : cqlSessionSpec.initQueries()) {
+        session.execute(query);
+      }
+    } else {
+      LOG.debug(
+          "Not creating session for {} because @{}.createSession() == false",
+          context.getElement(),
+          CqlSessionSpec.class.getSimpleName());
+    }
+  }
+
+  private void destroySession(ExtensionContext context) {
+    if (session != null) {
+      try {
+        if (cqlSessionSpec.dropKeyspace()) {
+          LOG.debug("Dropping keyspace {}", keyspaceId.asCql(true));
+          session.execute(String.format("DROP KEYSPACE IF EXISTS %s", keyspaceId.asCql(false)));
+        }
+      } finally {
+        LOG.debug("Destroying session for {}", context.getElement());
+        session.close();
+      }
+    }
+  }
+
+  private CqlSessionSpec getCqlSessionSpec(ExtensionContext context) {
+    AnnotatedElement element =
+        context
+            .getElement()
+            .orElseThrow(() -> new IllegalStateException("Expected to have an element"));
+    Optional<CqlSessionSpec> maybeSpec =
+        AnnotationSupport.findAnnotation(element, CqlSessionSpec.class);
+    if (!maybeSpec.isPresent() && element instanceof Method) {
+      maybeSpec =
+          AnnotationSupport.findAnnotation(
+              ((Method) element).getDeclaringClass(), CqlSessionSpec.class);
+    }
+    return maybeSpec.orElse(DefaultCqlSessionSpec.INSTANCE);
+  }
+
+  private CqlSessionBuilder newSessionBuilder(
+      StargateEnvironmentInfo stargate, ExtensionContext context) {
+    OptionsMap options = defaultConfig();
+    maybeCustomizeOptions(options, cqlSessionSpec, context);
+    CqlSessionBuilder builder = CqlSession.builder().withAuthCredentials("cassandra", "cassandra");
+    builder = maybeCustomizeBuilder(builder, cqlSessionSpec, context);
+    builder.withConfigLoader(DriverConfigLoader.fromMap(options));
+    for (StargateConnectionInfo node : stargate.nodes()) {
+      builder.addContactPoint(new InetSocketAddress(node.seedAddress(), node.cqlPort()));
+    }
+    return builder;
+  }
+
+  /**
+   * Waits until a CQL session sees all the Stargate nodes in its metadata. The CI environment is
+   * slow and it may take a while until system.peers is up to date.
+   *
+   * <p>Note: this is exposed publicly to be accessible from tests that create their own sessions.
+   */
+  public static void awaitAllNodes(CqlSession session, StargateEnvironmentInfo stargate) {
```

Comment:
Would do you think about adding an injected interface for this API? This way we can avoid test code calling static methods (any methods in fact) on the extension class.

Reply:
Right, that's more elegant, and not too complicated to add. I'll do it.